Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,196 @@
+import Foundation
+
+/// Corpus & Scheduler based on 
+/// Coverage-based Greybox Fuzzing as Markov Chain paper
+/// https://mboehme.github.io/paper/TSE18.pdf
+/// Simply put, the corpus keeps track of which paths have been found, and prioritizes seeds
+/// whose path has been hit less than average. Ideally, this allows the fuzzer to prioritize
+/// less explored coverage.
+
+public class MarkovCorpus: ComponentBase, Corpus {
+
+    // All programs currently in the corpus
+    private var allIncludedPrograms: [Program]
+    // Queue of programs to be executed next, all of which hit a rare edge
+    private var programExecutionQueue: [Program]
+
+    // For each edge encountered thus far, track which program initially discovered it
+    private var edgeMap: [UInt64:Program]
+
+    // This scheduler tracks the total number of samples it has returned
+    // This allows it to build an initial baseline by randomly selecting a program to mutate
+    // before switching to the more computationally expensive selection of programs that
+    // hit infreqent edges
+    private var totalExecs: UInt64
+
+    // This scheduler returns one base program multiple times, in order to compensate the overhead caused by tracking
+    // edge counts
+    private var currentProg: Program?
+    private var remainingEnergy: UInt64
+
+    // Markov corpus requires an evaluator that tracks edge coverage
+    // Thus, the corpus object keeps a reference to the evaluator, in order to only downcast once
+    private var covEvaluator: ProgramCoverageEvaluator
+
+    // This is required as MarkovCorpus increments all edges by this value when they are selected,
+    // to prevent issues with edge non-determinism
+    private var numConsecutiveMutations: UInt64
+
+    // This ensures the coverage map is done correctly for the initial program
+    private var genericProg: Program?
+
+    public init(numConsecutiveMutations: Int, evaluator: ProgramEvaluator) {
+        self.allIncludedPrograms = []
+        self.programExecutionQueue = []
+        self.totalExecs = 0
+        self.edgeMap = [:]
+        self.currentProg = nil
+        self.remainingEnergy = 0
+        self.numConsecutiveMutations = UInt64(numConsecutiveMutations)
+        if let covEvaluator = evaluator as? ProgramCoverageEvaluator {
+            self.covEvaluator = covEvaluator
+            covEvaluator.enableEdgeTracking()
+        } else {
+            // covEvaluator needs to be set prior to super.init, which means logger hasn't been instantiated
+            print("Markov corpus requires the use of a ProgramCoverageEvaluator as its evaluator")
```

Comment:
Would it be possible to change `evaluator: ProgramEvaluator` to `evaluator: ProgramCoverageEvaluator` in the signature to avoid this failure case entirely?

Reply:
And then maybe here or in `initialize` assert that `covEvaluator === fuzzer.evaluator`